### PR TITLE
[MM-44063] Line chart for top channels

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Line graph for top channels

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-44063

#### Screenshots
Screen Shot 2022-05-24 at 2 06 28 PM

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: components/activity_and_insights/insights/top_channels/top_channels_line_chart/top_channels_line_chart.tsx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        return {
+            labels,
+            values,
+        };
+    }, []);
+
+    const getGraphData = useMemo(() => {
+        const data = sortGraphData;
+        const dataset = [];
+
```
Comment (mkraft) [inline components/activity_and_insights/insights/top_channels/top_channels_line_chart/top_channels_line_chart.tsx:54]: What would you think about putting the colours in an array and then doing this in a loop?

```js
const colours = [theme.buttonBg, theme.onlineIndicator, theme.awayIndicator, theme.dndIndicator, theme.newMessageSeparator];

for (i = 0; i < 5; i++) {
        if (topChannels[i]) {
            const colour = colours[i];
            dataset.push({
                fillColor: colour,
                borderColor: colour,
                pointBackgroundColor: theme.buttonBg,
                pointBorderColor: colour,
                backgroundColor: 'transparent',
                pointRadius: 0,
                hoverBackgroundColor: colour,
                label: topChannels[i].display_name,
                data: data.values[topChannels[i].id],
                hitRadius: 10,
            });
        }
}
```

File: components/activity_and_insights/insights/top_channels/top_channels_line_chart/top_channels_line_chart.tsx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                            maxTicksLimit: 5,
+                            beginAtZero: true,
+                        },
+                    }],
+                },
+                tooltips: {
+                    callbacks: {
+                        label(tooltipItem, data) {
+                            const index = tooltipItem.datasetIndex;
+                            if (typeof index !== 'undefined' && data.datasets && data.datasets[index] && data.datasets[index].label) {
```
Comment (mkraft) [inline components/activity_and_insights/insights/top_channels/top_channels_line_chart/top_channels_line_chart.tsx:180]: I think this is equivalent...
```suggestion
                            if (typeof index !== 'undefined' && data?.datasets[index]?.label) {
```

**Review by mkraft — APPROVED** [review]

**Review by plant99 — APPROVED** [review]

**Comment by BenCookie95** [comment]
@matthewbirtch & @furqanmlk The line chart is available here https://mattermost-server-pr-20131.test.mattermost.cloud/

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Nice work @BenCookie95. It's coming along nicely. A few things I'm hoping we can tweak
1. Are we able style the tooltip to make a few adjustments?
    - update the font-family to use `Open Sans`
    - update the font-weight on the message count line to be '400' (normal) and change the opacity of the text color to 64% on this line
    - add 4px of space between the color square and the text label on the first line
    - Can we change the color square to a circle and remove the outline?
    - I think we might need to left-align everything since it looks quite off when multiple items are showing in the tooltip

image

2. Can we update the font used for the axis labels?
    - Should be using `Opens Sans` for the font-family
    - Can the font size for the labels be 10px?
    - Can we change the label color to use 72% opacity center channel color?

3. Are we able to modify the date format used on the x-axis? It's hard to read in this format YYYY-MM-DD. I'd like to see a friendlier format if possible:
    - If the date is within the current year, can we use Feb 15 (for example)
    - If the date is from a previous year, can we use Feb 15, 2021

4. Are we able to reduce the number of labels that show on the bottom axis any further? It can get a little hard to read at some screen sizes — particularly when there are a lot of hours shown for the day.
image

5. Can we make sure the time that displays follows the current user's timezone? It doesn't seem to be doing this right now since it's showing 16:00 as the most recent hour, but it's only 12:37 in my timezone.

6. Actually, we should also follow the user's preference for `Clock Display` (12-hour vs. 24-hour clock)

4. It seems that the tooltip gets the top cut off if there are a lot of items in it. I only see this when filtered on the `Last 28 days` 
image

8. The Top Channels card seems to flicker a height change when switching between date filters. Not a big deal, but I wondered if there was any way to smooth that out.

**Comment by BenCookie95** [comment]
@matthewbirtch I've fixed everything I can, some comments on the remaining issues below:

1. Are we able style the tooltip to make a few adjustments?
    * Can we change the color square to a circle and remove the outline? - _That isn’t possible unless I build a custom tooltip. This could be something I do in the future though_

4. Are we able to reduce the number of labels that show on the bottom axis any further? It can get a little hard to read at some screen sizes — particularly when there are a lot of hours shown for the day. _For `today` I have updated the labels to be every 2nd value if there are 8-12 labels and every 4th value if there are more than 12. Is that ok?_
5. Can we make sure the time that displays follows the current user's timezone? It doesn't seem to be doing this right now since it's showing 16:00 as the most recent hour, but it's only 12:37 in my timezone. - _Martin is looking into this on the backend. _

**Comment by matthewbirtch** [comment]
Thanks for resolving those items @BenCookie95. It's looking a lot better. A few more things I'm seeing now:
1. We should truncate the channel name in the tooltip. It makese the tooltip super-wide. 
image
2. This is unrelated to this PR, but we should create a follow-up to fix this: when I switch to other channels and come back to the Insights view, it's not remembering the filters that were previously applied (Team vs My Insights, and time period selection). We should keep that selection persistent. Right now it seems to reset to the default

**Comment by BenCookie95** [comment]
> Thanks for resolving those items @BenCookie95. It's looking a lot better. A few more things I'm seeing now:
> 
> 1. We should truncate the channel name in the tooltip. It makese the tooltip super-wide.
> 
> image
> 
> 2. This is unrelated to this PR, but we should create a follow-up to fix this: when I switch to other channels and come back to the Insights view, it's not remembering the filters that were previously applied (Team vs My Insights, and time period selection). We should keep that selection persistent. Right now it seems to reset to the default

I can fix the first point but number 2 would have to be addressed in a different ticket

**Comment by BenCookie95** [comment]
> Thanks for resolving those items @BenCookie95. It's looking a lot better. A few more things I'm seeing now:
> 
> 1. We should truncate the channel name in the tooltip. It makese the tooltip super-wide.
> 
> image
> 
> 2. This is unrelated to this PR, but we should create a follow-up to fix this: when I switch to other channels and come back to the Insights view, it's not remembering the filters that were previously applied (Team vs My Insights, and time period selection). We should keep that selection persistent. Right now it seems to reset to the default

@matthewbirtch I have truncated it at a max of 16 characters, let me know if you want that increased/decreased

**Comment by matthewbirtch** [comment]
> I can fix the first point but number 2 would have to be addressed in a different ticket

Can you create a ticket for this so it doesn't get list @BenCookie95?

**Review by matthewbirtch — APPROVED** [review]
Haven't seen it on the test server, but I'm good to approve this knowing you've made the necessary changes

**Review by furqanmlk — APPROVED** [review]
Line chart for top channels is working fine.

**Pull request merged**